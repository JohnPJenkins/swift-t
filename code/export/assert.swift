
// Assertions

#ifndef ASSERT_SWIFT
#define ASSERT_SWIFT

@assertion @builtin_op=ASSERT
assert(boolean condition, string msg) "turbine" "0.0.2" "assert";
@assertion @builtin_op=ASSERT_EQ
assertEqual(string|int|float|boolean v1, string|int|float|boolean v2,
               string msg) "turbine" "0.0.2" "assertEqual";
@assertion
assertLT(string|int|float|boolean v1, string|int|float|boolean v2,
               string msg) "turbine" "0.0.2" "assertLT";
@assertion
assertLTE(string|int|float|boolean v1, string|int|float|boolean v2,
               string msg) "turbine" "0.0.2" "assertLTE";

#endif
