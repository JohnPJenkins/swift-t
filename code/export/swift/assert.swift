
// Assertions

#ifndef ASSERT_SWIFT
#define ASSERT_SWIFT

assert(boolean condition, string msg) "turbine" "0.0.2" "assert";
assertEqual(string|int|float|boolean v1, string|int|float|boolean v2,
               string msg) "turbine" "0.0.2" "assertEqual";
assertLT(string|int|float|boolean v1, string|int|float|boolean v2,
               string msg) "turbine" "0.0.2" "assertLT";
assertLTE(string|int|float|boolean v1, string|int|float|boolean v2,
               string msg) "turbine" "0.0.2" "assertLTE";

#endif
