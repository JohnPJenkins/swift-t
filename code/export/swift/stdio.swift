#ifndef __STC_STDIO_SWIFT__
#define __STC_STDIO_SWIFT__

// STDIO.SWIFT

printf(string fmt, int|float|string|boolean... args) "turbine" "0.0.2" "printf";
#endif
