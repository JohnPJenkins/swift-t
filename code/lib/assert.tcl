# Turbine builtin assert functions for debugging

namespace eval turbine {

    # assert(cond, msg)
    # If cond is 0, then terminate program, printing msg
    proc assert { stack noresult inputs } {
        set cond [ lindex $inputs 0 ]
        set msg [ lindex $inputs 1 ]
        set rule_id [ rule_new ]
        rule $rule_id "assert-$cond-$msg" "$cond $msg" "" \
            "tf: assert_body $cond $msg"
    }
    
    proc assert_body { cond msg } {
        set cond_value [ get $cond ]
        set msg_value [ get $msg ]
        assert_impl $cond_value $msg_value
    }

    proc assert_impl { cond msg } {
        if { $cond == 0 } {
            error "Assertion failed!: $msg"
        } else {
            log "assert passed: $msg"
        }
    }
    
    # assert(arg1, arg2, msg)
    # if arg1 != arg2 according to TCL's comparison rules
    #  (e.g. "1" == 1 and 1 == 1)
    # then crash, printing the two values and the provided msg
    proc assertEqual { stack noresult inputs } {
        set arg1 [ lindex $inputs 0 ]
        set arg2 [ lindex $inputs 1 ]
        set msg [ lindex $inputs 2 ]
        set rule_id [ rule_new ]
        rule $rule_id "assertEqual-$arg1-$arg2-$msg" "$arg1 $arg2 $msg" "" \
            "tf: assertEqual_body $arg1 $arg2 $msg"
    }
    
    proc assertEqual_body { arg1 arg2 msg } {
        set arg1_value [ get $arg1 ]
        set arg2_value [ get $arg2 ]
        set msg_value [ get $msg ]
        assertEqual_impl $arg1_value $arg2_value $msg_value
    }

    proc assertEqual_impl { arg1 arg2 msg } {
        if { $arg1 != $arg2 } {
            error "Assertion failed $arg1 != $arg2: $msg"
        } else {
            log "assertEqual $arg1 $arg2 passed: $msg"
        }
    }
    
    # assertLT(arg1, arg2, msg)
    # if arg1 >= arg2 according to TCL's comparison rules
    # then crash, printing the two values and the provided msg
    proc assertLT { stack noresult inputs } {
        set arg1 [ lindex $inputs 0 ]
        set arg2 [ lindex $inputs 1 ]
        set msg [ lindex $inputs 2 ]
        set rule_id [ rule_new ]
        rule $rule_id "assertLT-$arg1-$arg2-$msg" "$arg1 $arg2 $msg" "" \
            "tf: assertLT_body $arg1 $arg2 $msg"
    }
    
    proc assertLT_body { arg1 arg2 msg } {
        set arg1_value [ get $arg1 ]
        set arg2_value [ get $arg2 ]
        if { $arg1_value >= $arg2_value } {
            set msg_value [ get $msg ]
            error "Assertion failed $arg1_value >= $arg2_value: $msg_value"
        } else {
            log "assertLT: $arg1_value $arg2_value"
        }
    }
    
    # assertLTE(arg1, arg2, msg)
    # if arg1 >= arg2 according to TCL's comparison rules
    # then crash, printing the two values and the provided msg
    proc assertLTE { stack noresult inputs } {
        set arg1 [ lindex $inputs 0 ]
        set arg2 [ lindex $inputs 1 ]
        set msg [ lindex $inputs 2 ]
        set rule_id [ rule_new ]
        rule $rule_id "assertLTE-$arg1-$arg2-$msg" "$arg1 $arg2 $msg" "" \
            "tf: assertLTE_body $arg1 $arg2 $msg"
    }
    
    proc assertLTE_body { arg1 arg2 msg } {
        set arg1_value [ get $arg1 ]
        set arg2_value [ get $arg2 ]
        if { $arg1_value > $arg2_value } {
            set msg_value [ get $msg ]
            error "Assertion failed $arg1_value > $arg2_value: $msg_value"
        } else {
            log "assertLTE: $arg1_value $arg2_value"
        }
    }
}
