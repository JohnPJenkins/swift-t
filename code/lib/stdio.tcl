
# I/O library routines

namespace eval turbine {

    # namespace export printf

    proc printf { args } {

        set a [ lindex $args 2 ]
        rule printf $a $turbine::LOCAL \
            "printf_body $a"
    }
    proc printf_body { args } {
        set L [ list ]
        foreach a $args {
            lappend L [ retrieve $a ]
        }
        set s [ eval format $L ]
        puts $s
    }

}
