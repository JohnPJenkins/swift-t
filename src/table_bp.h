/*
 * Copyright 2013 University of Chicago and Argonne National Laboratory
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License
 */

/*
  Table which has binary key as key
 */
#ifndef __TABLE_BP_H
#define __TABLE_BP_H

#include <stdbool.h>
#include <stddef.h>

#include "binkeys.h"

typedef struct table_bp_entry table_bp_entry;

struct table_bp_entry
{
  // NULL key indicates empty entry.  Entry entries should only be used
  // to mark empty hash table buckets
  void* key;
  size_t key_len;
  void* data; // NULL is valid data
  table_bp_entry *next;
};

typedef struct table_bp
{
  table_bp_entry *array;
  int capacity;
  int size;
  float load_factor;
  int resize_threshold; // Resize if > this size
} table_bp;

#define TABLE_BP_DEFAULT_LOAD_FACTOR 0.75

/*
  Macro for iterating over table entries.  This handles the simple case
  of iterating over all valid table entries with no modifications.
 */
#define TABLE_BP_FOREACH(T, item) \
  for (int __i = 0; __i < T->capacity; __i++) \
    if (table_bp_entry_valid(&T->array[__i])) \
      for (table_bp_entry *item = &T->array[__i]; item != NULL; \
           item = item->next)

bool table_bp_init(table_bp* target, int capacity);

bool table_bp_init_custom(table_bp* target, int capacity, float load_factor);

table_bp* table_bp_create(int capacity);

table_bp*
table_bp_create_custom(int capacity, float load_factor);

bool table_bp_add(table_bp *target, const void* key,
                  size_t key_len, void* data);

bool table_bp_set(table_bp* target, const void* key,
                  size_t key_len, void* value, void** old_value);

bool table_bp_search(const table_bp* target, const void* key,
                  size_t key_len, void **value);

bool table_bp_contains(const table_bp* table, const void* key,
                  size_t key_len);

bool table_bp_remove(table_bp* table, const void* key,
                  size_t key_len, void** data);

void table_bp_dump(const char* format, const table_bp* target);

/*
  Free data structure, and callback function with key and value
 */
void table_bp_free_callback(table_bp* target, bool free_root,
                            void (*callback)(void*, size_t, void*));

void table_bp_free(table_bp* target);

void table_bp_destroy(table_bp* target);

void table_bp_release(table_bp* target);

size_t table_bp_keys_string(char** result,
                            const table_bp* target);

size_t table_bp_keys_string_slice(char** result,
                            const table_bp* target,
                            int count, int offset);

size_t table_bp_keys_tostring(char* result,
                              const table_bp* target);

size_t table_bp_keys_tostring_slice(char* result,
                              const table_bp* target,
                              int count, int offset);

void  table_bp_dumpkeys(const table_bp* target);

/*
  If the entry contains data
 */
static inline bool
table_bp_entry_valid(const table_bp_entry *e)
{
  return e->key != NULL;
}

/*
  Check if key matches item key. Inline for performance
 */
static inline bool
table_bp_key_match(const void *key, size_t key_len, const table_bp_entry *e)
{
  return table_bp_entry_valid(e) && bin_key_eq(key, key_len, e->key, e->key_len);
}
#endif // __TABLE_BP_H
