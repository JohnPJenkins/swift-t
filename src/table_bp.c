/*
 * Copyright 2013 University of Chicago and Argonne National Laboratory
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License
 */

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "binkeys.h"

#include "table_bp.h"
#include "c-utils-types.h"


// Double in size for now
static const float table_bp_expand_factor = 2.0;

static void
table_bp_dump2(const char* format, const table_bp* target, bool include_vals);

static bool table_bp_item_add_head(table_bp_item *head,
            void* key, size_t key_len, void* data);

static bool table_bp_item_add_tail(table_bp_item *head,
            table_bp_item *entry, bool copy_entry);

static void
table_bp_remove_entry(table_bp_entry *e, table_bp_entry *prev);

static bool
table_bp_expand(table_bp *T);

static void init_entry(table_bp_entry *entry)
{
  entry->key = NULL;
  entry->key_len = 0;
  entry->data = NULL;
  entry->next = NULL;
}

static void
update_resize_threshold(table_bp *T)
{
  T->resize_threshold = (int)(capacity * load_factor);
}

/**
   
*/
bool
table_bp_init(table_bp* target, int capacity, float load_factor)
{
  assert(capacity >= 1);
  target->size = 0;
  target->capacity = capacity;
  target->load_factor = load_factor;
  update_resize_threshold(&target);

  target->array = malloc(sizeof(target->array[0]) * (size_t)capacity);
  if (!target->array)
  {
    return false;
  }

  for (int i = 0; i < capacity; i++)
  {
    init_entry(&target->array[i]);
  }
  return true;
}

table_bp*
table_bp_create(int capacity)
{
  table_bp* new_table =  malloc(sizeof(const table_bp));
  if (! new_table)
    return NULL;

  bool result = table_bp_init(new_table, capacity);
  if (!result)
  {
    free(new_table);
    return NULL;
  }

  return new_table;
}

void
table_bp_free(table_bp* target)
{
  table_bp_free_callback(target, true, NULL);
}

void table_bp_free_callback(table_bp* target, bool free_root,
                         void (*callback)(void*, size_t, void*))
{
  for (int i = 0; i < target->capacity; i++)
  {
    bool head = true;
    for (table_bp_entry *e = &target->array[i]; e != NULL; e = e->next)
    {
      if (e->key != NULL && callback != NULL)
        callback(e->key, e->key_len, e->data);

      if (!head) {
        // head is stored in array, rest were malloced separately
        free(e);
      }
      head = false;
    }
  }

  free(target->array);

  if (free_root)
  {
    free(target);
  }
  else
  {
    target->array = NULL;
    target->capacity = target->size = 0;
  }
}

void
table_bp_destroy(table_bp* target)
{
  for (int i = 0; i < target->capacity; i++)
  {
    bool head = true;
    for (table_bp_entry *e = &target->array[i]; e != NULL; e = e->next)
    {
      if (e->key != NULL)
      {
        // Entry was valid
        free(item->key);
        free(item->data);
        item->key = item->data = NULL;
      }
      if (!head)
      {
       free(e);
      }
    }
  }

  free(target->array);
  free(target);
}

void
table_bp_release(table_bp* target)
{
  free(target->array);
}


static bool table_bp_item_add_head(table_bp_item *head,
            void* key, size_t key_len, void* data)
{
  if (table_bp_entry_valid(head))
  {
    // Head occupied: replace head with this one
    table_bp_entry *e = malloc(sizeof(table_bp_entry));
    if (e == NULL)
    {
      return false;
    }
    // Copy current head to e
    memcpy(e, head, sizeof(*e));
    head->next = e;
  }

  // Put new data in head
  head->key = key;
  head->key_len = key_len;
  head->data = data;
  return true;
}

/**
  Add at end of bucket chain
  copy_entry: if true, allocate new entry,
              otherwise reuse provided entry or free
              
 */
static bool table_bp_item_add_tail(table_bp_item *head,
            table_bp_item *entry, bool copy_entry)
{
  if (table_bp_entry_valid(head))
  {
    // Head occupied: find tail and add
    if (copy_entry)
    {
      table_bp_item *new_entry = malloc(sizeof(*entry));
      if (new_entry == NULL)
      {
        return false;
      }
      memcpy(new_entry, entry, sizeof(*entry));
      entry = new_entry;
    }

    table_bp_entry *tail = head; // Find tail of list
    while (tail->next != NULL)
    {
      tail = tail->next;
    }
    tail->next = entry;
    entry->next = NULL;
  }
  else
  {
    assert(head->next == NULL); // Check no dangling
    // Empty list case
    memcpy(head, e, sizeof(*head));
    head->next = NULL;

    if (!copy_entry)
    {
      // Don't need memory since we copied to head
      free(e);
    }
  }

  return true;
}

/**
   Note: duplicates internal copy of key (in list_bp_add())
 */
bool
table_bp_add(table_bp *target, const void* key, size_t key_len,
             void* data)
{
  // Check to resize hash table
  if (target->size >= target->resize_threshold)
  {
    bool ok = table_bp_expand(target);
    if (!ok)
      return false;
  }
  int index = bin_key_hash(key, key_len, target->capacity);

  /* 
   * Add at head of list to avoid traversing list.  This means that in
   * case of duplicate keys, the newest is returned
   */
  table_bp_entry *head = &target->array[index];


  void *key_copy = malloc(key_len);
  if (key_copy == NULL)
  {
    return false;
  }
  memcpy(key_copy, key, key_len);

  bool ok = table_bp_item_add_head(head, key_copy, key_len, data);
  if (ok)
  {
    target->size++;
    return true;
  }
  else
  {
    free(key_copy);
    return false;
  }
}

/**
   If found, caller is responsible for old_value -
          it was provided by the user
   @return True if found
 */
bool
table_bp_set(table_bp* target, const void* key, size_t key_len,
          void* value, void** old_value)
{
  int index = bin_key_hash(key, key_len, target->capacity);

  for (table_bp_entry *e = &target->array[index]; e != NULL; e = e->next)
  {
    if (table_bp_key_match(key, key_len, e))
    {
      *old_value = e->data;
      e->data = value;
      return true;
    }
  }

  return false;
}

/**
   @param value: this is used to return the value if found
   @return true if found, false if not 
   
   if value is NULL and return is true, this means that the key exists
   but the value is NULL
*/
bool
table_bp_search(const table_bp* table, const void* key,
                size_t key_len, void **value)
{
  int index = bin_key_hash(key, key_len, table->capacity);

  for (table_bp_entry *e = &target->array[index]; e != NULL; e = e->next)
  {
    if (table_bp_key_match(key, key_len, e)) {
      *value = e->data;
      return true;
    }

  *value = NULL;
  return false;
}

bool
table_bp_contains(const table_bp* table, const void* key,
                  size_t key_len)
{
  void* tmp = NULL;
  return table_bp_search(table, key, key_len, &tmp);
}

/*
 Unlink and free entry if needed
 prev: previous entry, or NULL if list head
 */
static void
table_bp_remove_entry(table_bp_entry *e, table_bp_entry *prev)
{
  if (prev == NULL)
  {
    // Removing head of list
    if (e->next == NULL)
    {
      // List is now empty - reset
      init_entry(e);
    }
    else
    {
      // Promote other entry to head
      table_bp_entry *next = e->next;
      memcpy(e, next, sizeof(*e));
      free(next);
    }
  }
  else
  {
    prev->next = e->next;
    free(e); // Was malloced separately
  }
}

bool
table_bp_remove(table_bp* table, const void* key, size_t key_len,
                void** data)
{
  int index = bin_key_hash(key, key_len, table->capacity);
  table_bp_entry *head = &target->array[index];
  table_bp_entry *prev = NULL;
  for (table_bp_entry *e = head; e != NULL; e = e->next)
  {
    if (table_bp_key_match(key, key_len, e)) {
      *data = e->data; // Store data for caller
      free(e->key);

      table_bp_remove_entry(e, prev); 
      table->size--;
      return true;
    }
    prev = e;
  }
  return false;
}

/**
   Resize hash table to be larger
 */
static bool
table_bp_expand(table_bp *T)
{
  int new_capacity = table_bp_expand_factor * T->capacity;
  assert(new_capacity > T->capacity);
  table_bp_entry *new_array = malloc(sizeof(T->array[0]) *
                                     (size_t)new_capacity);

  if (new_array == NULL)
  {
    return false;
  }
  
  
  for (int i = 0; i < new_capacity; i++)
  {
    init_entry(&new_array[i]);
  }
  
  // Rehash and move all entries from old table
  for (int i = 0; i < target->capacity; i++)
  {
    table_bp_entry *head = &target->array[i];
    if (!table_bp_entry_valid(head))
    {
      assert(head->next == NULL);
      continue; // Bucket was empty
    }

    bool is_head = true;
    for (table_bp_entry *e = head; e != NULL; e = e->next)
    {
      // all entries should be valid if we get here
      assert(table_bp_entry_valid(e));
      int new_ix = bin_key_hash(key, key_len, new_capacity);
      bool ok = table_bp_item_add_tail(&new_array[new_ix], e, is_head);
      if (!ok)
      {
        // TODO: how to handle? we're in an awkward state in the middle
        //  of moving entries.
        //  It's unlikely that this will happen.  It is impossible if
        //  we're doubling the size of the array since we don't need to
        //  allocate any memory then.
        return false;
      }

      is_head = false;
    }
  }
  
  free(T->array);
  T->array = new_array;
  T->capacity = new_capacity;
  update_resize_threshold(T); 
  return true;
}

/** format specifies the output format for the data items
  TODO: format ignored
 */
void
table_bp_dump(const char* format, const table_bp* target)
{
  table_bp_dump2(format, target, true);
}

void
table_bp_dumpkeys(const table_bp* target)
{
  table_bp_dump2(NULL, target, false);
}

static void
table_bp_dump2(const char *format, const table_bp* target, bool include_vals)
  printf("{\n");
  for (int i = 0; i < target->capacity; i++)
  {
    table_bp_entry *head = &target->array[i];
    if (!table_bp_entry_valid(head) && head->next == NULL)
    {
      // Skip empty buckets
      continue;
    }
    printf("%i: ", i);

    for (table_bp_entry *e = head; e != NULL; e = e->next)
    {
      printf("(");
      printf_key(e->key, e->key_len);
      if (include_vals)
      {
        printf(", %s)", (char*) data);
      }
      else
      {
        printf(")");
      }
      if (e->next != NULL)
        printf(", ");
    }
    printf("\n");
  }
  printf("}\n");
}


size_t
table_bp_keys_string_length(const table_bp* target)
{
  size_t result = 0;
  for (int i = 0; i < target->capacity; i++)
    result += list_bp_keys_string_length(target->array[i]);
  return result;
}

size_t
table_bp_keys_string(char** result, const table_bp* target)
{
  size_t length = table_bp_keys_string_length(target);
  // Allocate size for each key and a space after each one
  *result = malloc(length + (size_t)(target->size+1));
  // Update length with actual usage
  length = table_bp_keys_tostring(*result, target);
  return length;
}

size_t
table_bp_keys_string_slice(char** result,
                        const table_bp* target,
                        int count, int offset)
{
  size_t length = table_bp_keys_string_length(target);
  // Allocate size for each key and a space after each one
  *result = malloc(length + (size_t)(target->size+1));
  // Update length with actual usage
  length = table_bp_keys_tostring_slice(*result, target, count, offset);
  return length;
}

size_t
table_bp_keys_tostring(char* result, const table_bp* target)
{
  char* p = result;
  for (int i = 0; i < target->capacity; i++)
    p += list_bp_keys_tostring(p, target->array[i]);
  return (size_t)(p-result);
}

size_t
table_bp_keys_tostring_slice(char* result, const table_bp* target,
                          int count, int offset)
{
  // Count of how many items we have covered
  int c = 0;
  char* p = result;
  p[0] = '\0';
  for (int i = 0; i < target->capacity; i++)
  {
    struct list_bp* L = target->array[i];
    for (struct list_bp_item* item = L->head; item; item = item->next)
    {
      if (c < offset) {
        c++;
        continue;
      }
      if (c >= offset+count && count != -1)
        break;
      p += sprintf_key(p, item->key, item->key_len);
      *(p++) = ' ';
      c++;
    }
  }
  return (size_t)(p-result);
}

/** Dump list_bp to string as in snprintf()
    size must be greater than 2.
    format specifies the output format for the data items
    internally allocates O(size) memory
    returns int greater than size if size limits are exceeded
            indicating result is garbage
*/
size_t
table_bp_tostring(char* output, size_t size,
               char* format, const table_bp* target)
{
  size_t   error = size+1;
  char* ptr   = output;
  int i;
  ptr += sprintf(output, "{\n");

  char* s = malloc(sizeof(char) * size);

  for (i = 0; i < target->capacity; i++)
  {
    size_t r = list_bp_tostring(s, size, format, target->array[i]);
    if (((size_t)(ptr-output)) + r + 2 < size)
      ptr += sprintf(ptr, "%s\n", s);
    else
      return error;
  }
  sprintf(ptr, "}\n");

  free(s);
  return (size_t)(ptr-output);
}
