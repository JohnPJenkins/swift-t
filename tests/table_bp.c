/*
 * Copyright 2013 University of Chicago and Argonne National Laboratory
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License
 */

#include <stdio.h>

#include <table_bp.h>
#include <assert.h>

int main() {

  table_bp T;
  bool ok;

  ok = table_bp_init(&T, 4);
  assert(ok);

  // force expand
  for (int i = 0; i < 16; i++)
  {
    char key[128];
    size_t key_len = (size_t) sprintf(key, "key%i", i);
    
    // remove null terminator, since it shouldn't depend on it
    key[key_len] = ' ';

    ok = table_bp_add(&T, key, key_len, NULL);
    assert(ok);

    // Check iteration works;
    int count = 0;
    TABLE_BP_FOREACH(&T, item)
    {
      count++;
    }
    printf("i=%i count=%i\n", i, count);
    assert(count == i + 1);
  }


  printf("DONE\n");
}
