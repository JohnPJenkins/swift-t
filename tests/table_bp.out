i=0 count=1
i=1 count=2
i=2 count=3
i=3 count=4
i=4 count=5
i=5 count=6
i=6 count=7
i=7 count=8
i=8 count=9
i=9 count=10
i=10 count=11
i=11 count=12
i=12 count=13
i=13 count=14
i=14 count=15
i=15 count=16
i=16 count=17
i=17 count=18
i=18 count=19
i=19 count=20
i=20 count=21
i=21 count=22
i=22 count=23
i=23 count=24
i=24 count=25
i=25 count=26
i=26 count=27
i=27 count=28
i=28 count=29
i=29 count=30
i=30 count=31
i=31 count=32
i=32 count=33
i=33 count=34
i=34 count=35
i=35 count=36
i=36 count=37
i=37 count=38
i=38 count=39
i=39 count=40
i=40 count=41
i=41 count=42
i=42 count=43
i=43 count=44
i=44 count=45
i=45 count=46
i=46 count=47
i=47 count=48
i=48 count=49
i=49 count=50
i=50 count=51
i=51 count=52
i=52 count=53
i=53 count=54
i=54 count=55
i=55 count=56
i=56 count=57
i=57 count=58
i=58 count=59
i=59 count=60
i=60 count=61
i=61 count=62
i=62 count=63
i=63 count=64
Search: key30=30
Search: key59=59
Search: key24=24
Search: key53=53
Search: key18=18
Search: key47=47
Search: key12=12
Search: key41=41
Search: key6=6
Search: key35=35
Search: key0=0
Search: key29=29
Search: key58=58
Search: key23=23
Search: key52=52
Search: key17=17
Search: key46=46
Search: key11=11
Search: key40=40
Search: key5=5
Search: key34=34
Search: key63=63
Search: key28=28
Search: key57=57
Search: key22=22
Search: key51=51
Search: key16=16
Search: key45=45
Search: key10=10
Search: key39=39
Search: key4=4
Search: key33=33
Search: key62=62
Search: key27=27
Search: key56=56
Search: key21=21
Search: key50=50
Search: key15=15
Search: key44=44
Search: key9=9
Search: key38=38
Search: key3=3
Search: key32=32
Search: key61=61
Search: key26=26
Search: key55=55
Search: key20=20
Search: key49=49
Search: key14=14
Search: key43=43
Search: key8=8
Search: key37=37
Search: key2=2
Search: key31=31
Search: key60=60
Search: key25=25
Search: key54=54
Search: key19=19
Search: key48=48
Search: key13=13
Search: key42=42
Search: key7=7
Search: key36=36
Search: key1=1
Remove: key5=5
Remove: key34=34
Remove: key63=63
Remove: key28=28
Remove: key57=57
Remove: key22=22
Remove: key51=51
Remove: key16=16
Remove: key45=45
Remove: key10=10
Remove: key39=39
Remove: key4=4
Remove: key33=33
Remove: key62=62
Remove: key27=27
Remove: key56=56
Remove: key21=21
Remove: key50=50
Remove: key15=15
Remove: key44=44
Remove: key9=9
Remove: key38=38
Remove: key3=3
Remove: key32=32
Remove: key61=61
Remove: key26=26
Remove: key55=55
Remove: key20=20
Remove: key49=49
Remove: key14=14
Remove: key43=43
Remove: key8=8
Remove: key37=37
Remove: key2=2
Remove: key31=31
Remove: key60=60
Remove: key25=25
Remove: key54=54
Remove: key19=19
Remove: key48=48
Remove: key13=13
Remove: key42=42
Remove: key7=7
Remove: key36=36
Remove: key1=1
Remove: key30=30
Remove: key59=59
Remove: key24=24
Remove: key53=53
Remove: key18=18
Remove: key47=47
Remove: key12=12
Remove: key41=41
Remove: key6=6
Remove: key35=35
Remove: key0=0
Remove: key29=29
Remove: key58=58
Remove: key23=23
Remove: key52=52
Remove: key17=17
Remove: key46=46
Remove: key11=11
Remove: key40=40
DONE
